pytorch_lightning/utilities/cli.py

```diff
@@ -0,0 +1,245 @@
+
+        self.init_parser()
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.before_parse_arguments(self.parser)
+        self.parse_arguments()
+        self.before_instantiate_classes()
+        self.instantiate_classes()
+        self.prepare_fit_kwargs()
+        self.before_fit()
+        self.fit()
+        self.after_fit()
+
+    def init_parser(self):
+        """Method that instantiates the argument parser"""
+        self.parser = LightningArgumentParser(**self.parser_kwargs)
+
+    def add_arguments_to_parser(self, parser: LightningArgumentParser):
+        """Implement to add extra arguments to parser
+
+        Args:
+            parser: The argument parser object to which arguments should be added
+        """
+        pass
+
+    def add_core_arguments_to_parser(self):
+        """Adds arguments from the core classes to the parser"""
+        self.parser.add_lightning_class_args(self.trainer_class, 'trainer')
+        self.parser.add_lightning_class_args(self.model_class, 'model', subclass_mode=self.subclass_mode_model)
+        if self.datamodule_class is not None:
+            self.parser.add_lightning_class_args(self.datamodule_class, 'data', subclass_mode=self.subclass_mode_data)
+
+    def before_parse_arguments(self, parser: LightningArgumentParser):
+        """Implement to run some code before parsing arguments
+
+        Args:
+            parser: The argument parser object that will be used to parse
+        """
+        pass
+
+    def parse_arguments(self):
+        """Parses command line arguments and stores it in self.config"""
+        self.config = self.parser.parse_args()
+
+    def before_instantiate_classes(self):
+        """Implement to run some code before instantiating the classes"""
+        pass
+
+    def instantiate_classes(self):
+        """Instantiates the classes using settings from self.config"""
+        self.config_init = self.parser.instantiate_subclasses(self.config)
+        self.instantiate_datamodule()
+        self.instantiate_model()
+        self.instantiate_trainer()
+
+    def instantiate_datamodule(self):
+        """Instantiates the datamodule using self.config_init['data'] if given"""
+        if self.datamodule_class is None:
+            self.datamodule = None
+        elif self.subclass_mode_data:
+            self.datamodule = self.config_init['data']
+        else:
+            self.datamodule = self.datamodule_class(**self.config_init.get('data', {}))
+
+    def instantiate_model(self):
+        """Instantiates the model using self.config_init['model']"""
+        if self.subclass_mode_model:
+            self.model = self.config_init['model']
+        else:
+            self.model = self.model_class(**self.config_init.get('model', {}))
+
+    def instantiate_trainer(self):
+        """Instantiates the trainer using self.config_init['trainer']"""
+        self.trainer_kwargs.update(self.config_init['trainer'])
```

Comment:
This seems to override any arguments passed to `trainer_kwargs`, thus making it useless as `self.config_init['trainer']` will contain all trainer arguments

Reply:
True. Will change it.

Reply:
I have thought about possible options. It does not make sense that `trainer_kwargs` given to `LightningCLI` have precedence over what is given in a config file or command line. And it would be very confusing if `trainer_kwargs` would only work for some, e.g. override only if default value is `None`. But the current `trainer_kwargs` is useful to set some `callbacks` so that they always available without need to add them in a config file. So I can think of only two sensible solutions. Remove `trainer_kwargs` or change it to be `trainer_callbacks`. If we remove it, then to add persistent callbacks would require to extend a method. Maybe changing to `trainer_callbacks` might be preferable since I guess this can be quite common. What do you think?

Reply:
> It does not make sense that trainer_kwargs given to LightningCLI have precedence over what is given in a config file or command line.

I agree

But it seems that `trainer_kwargs` gets overriden even if you provide no config/command-line arguments, for example:

```python
from pytorch_lightning.utilities.cli import LightningCLI
from tests.helpers.boring_model import BoringDataModule, BoringModel

cli = LightningCLI(
    BoringModel, BoringDataModule,
    trainer_kwargs={'fast_dev_run': True}
)
```

This will not run `fast_dev_run`

Why don't we write `self.config_init['trainer']` only if there is a config?

Reply:
Okay, so from what I am understanding it would work like overriding the defaults of the parser and this would only be overridden if given in a config file or command line. For this to work with the correct behavior it needs to be done a bit differently. Not by modifying a kwargs dict. But might be simple anyway.